### DISCUSSION TITLE
Make CMS backend area extensible

### DISCUSSION THREAD

**Discussion by Flynsarmy** [wintercms/winter#487] [Ideas]
Many plugins would benefit from integrating into the CMS area of backend instead of cluttering the main top menu bar. For example a plugin that manages carousels or menus. Extending the CMS area is currently impossible however as the it was not written with extensibility in mind.

The *modules/cms/controllers/Index.php* file has each base types (Pages, Partials, Layouts, etc) hardcoded into it. I suggest moving the registration of CMS types into their respective classes (*Cms\Classes\Page* etc) and implement an Interface for the required methods for any class wanting to extend CMS.

This approach has a few benefits.
* It will declutter the *Cms\Controllers\Index* controller and remove many of the dependencies pulled in at the top.
* It will force the CMS controller to be written in an extensible manner.
* It will provide a more centralised location for administrators to access CMS related content.
* It will provide code examples to third party plugins wanting to extend the CMS area.

Because no plugins currently extend the CMS (as it's not possible) there should be no backwards compatibility concerns here. If we can come up with a concrete plan for how we want this idea to be implemented, I'm happy to work on it myself.

**Comment by LukeTowers** [comment]
While I'm not opposed to this in theory, I see the CMS section becoming more of a "Theme Editor" section with a heavier emphasis on a code editor / IDE experience.

This could definitely be a good enhancement for the Winter.Pages plugin though; that is really the end user focused content management experience for Winter CMS. Thoughts?

**Comment by mjauvin** [comment]
Definitely agree for the Pages plugin, not sure exactly how that would prevent a rewrite of the CMS section to be more extensible.

**Reply by LukeTowers** [reply]
I wouldn't want to encourage people to be putting client friendly content management pieces like sliders etc into the CMS section. We could make it more extensible as a part of its general improvements, but I would want it to be more focused on the developer IDE experience, more file driven. 

Calling it "CMS" in the menu is a mistake IMHO and I don't want to encourage people to extend it with functionality that won't make sense to be there once we make the improvements that I want to.

**Comment by Flynsarmy** [comment]
Since we seem to be in agreement that CMS should be made extensible, we need to work up a game plan on how to make this happen. We currently have this sidemenu setup in CMS, Winter.Pages, Winter.Builder. These should all be standardised to use either a new Controller or a new ControllerBehavior.

I'm thinking a good first step would be in RainLab.Pages. I'll see how much I can break out of the Index controller and move into the content/menu/page/snippet Widgets.

**Reply by bennothommo** [reply]
@Flynsarmy the Pages plugin is probably the easiest to work off, but I believe the Builder plugin is the most "recent" implementation of this methodology out of the three. I'm working on a shop plugin at the moment based off the Winter.Pages implementation, but there's some neat tricks in the Builder plugin's implementation.

**Reply by Flynsarmy** [reply]
I'll take a look. Thanks!
